Compilation error when using assignment operator on a map element that has another map element as key

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** 0.2.4 6079448
**OS:** windows, Microsoft Windows 11 Pro v22000 64-bit

**What did you do?**
Compiled this code
```
m := {'a': 'x'}
mut n := {'x': 1}
n[m['a']] += 2
```

**What did you expect to see?**
Successful compilation
 
**What did you see instead?**
Error when compiling with `v -cg`: (tcc)
```
C:/Users/SL/AppData/Local/Temp/v_0/try-maps.9791927726383572084.tmp.c:455: warning: WINVER redefined
C:/Users/SL/AppData/Local/Temp/v_0/try-maps.9791927726383572084.tmp.c:5955: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/SL/AppData/Local/Temp/v_0/try-maps.9791927726383572084.tmp.c:6552: warning: assignment from incompatible pointer type
C:/Users/SL/AppData/Local/Temp/v_0/try-maps.9791927726383572084.tmp.c:6552: warning: assignment makes pointer from integer without a cast
C:/Users/SL/AppData/Local/Temp/v_0/try-maps.9791927726383572084.tmp.c:6552: warning: cast between pointer and integer of different size
C:/Users/SL/AppData/Local/Temp/v_0/try-maps.9791927726383572084.tmp.c:10831: error: ',' expected (got ")")
builder error:
==================
C error. This should never happen.
```

Error when compiling with `v -prod -cg`: (gcc)
```
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c: In function 'winapi_lasterr_str':
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:6510:195: warning: passing argument 5 of 'FormatMessageW' from incompatible pointer type [-Wincompatible-pointer-types]
 6510 |         voidptr res = FormatMessage(((FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM) | FORMAT_MESSAGE_IGNORE_INSERTS), NULL, err_msg_id, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), &msgbuf, 0, NULL);
      |
                                                              ^~~~~~~
      |
                                                              |
      |
                                                              u16 ** {aka short unsigned int **}
In file included from k:\scoop\apps\gcc\current\x86_64-w64-mingw32\include\windows.h:70,
                 from C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:470:
k:\scoop\apps\gcc\current\x86_64-w64-mingw32\include\winbase.h:1385:122: note: expected 'LPWSTR' {aka 'short unsigned int *'} but argument is of type 'u16 **' {aka 'short unsigned int **'}
 1385 |   WINBASEAPI DWORD WINAPI FormatMessageW (DWORD dwFlags, LPCVOID lpSource, DWORD dwMessageId, DWORD dwLanguageId, LPWSTR lpBuffer, DWORD nSize, va_list *Arguments);
      |                                                                                                                   ~~~~~~~^~~~~~~~
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:6510:23: warning: initialization of 'voidptr' {aka 'void *'} from 'DWORD' {aka 'long unsigned int'} makes pointer from integer without a cast [-Wint-conversion]
 6510 |         voidptr res = FormatMessage(((FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM) | FORMAT_MESSAGE_IGNORE_INSERTS), NULL, err_msg_id, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), &msgbuf, 0, NULL);
      |                       ^~~~~~~~~~~~~
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c: In function 'main__main':
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:10789:56: error: too many arguments to function 'map_set'
10789 |         (*((int*)map_get_and_set((map*)&n, &(string[]){map_set(&m, &(string[]){_SLIT("a")}, &(string[]) { }, &(int[]) { 0 })))2 });
      |                                                        ^~~~~~~
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:7514:6: note: declared here
 7514 | void map_set(map* m, voidptr key, voidptr value) {
      |      ^~~~~~~
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:10789:56: error: void value not ignored as it ought to be
10789 |         (*((int*)map_get_and_set((map*)&n, &(string[]){map_set(&m, &(string[]){_SLIT("a")}, &(string[]) { }, &(int[]) { 0 })))2 });
      |                                                        ^~~~~~~
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:10789:125: error: expected '}' before ')' token
10789 |         (*((int*)map_get_and_set((map*)&n, &(string[]){map_set(&m, &(string[]){_SLIT("a")}, &(string[]) { }, &(int[]) { 0 })))2 });
      |                                                       ~                                                                     ^
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:10789:18: error: too few arguments to function 'map_get_and_set'
10789 |         (*((int*)map_get_and_set((map*)&n, &(string[]){map_set(&m, &(string[]){_SLIT("a")}, &(string[]) { }, &(int[]) { 0 })))2 });
      |                  ^~~~~~~~~~~~~~~
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:7608:9: note: declared here
 7608 | voidptr map_get_and_set(map* m, voidptr key, voidptr zero) {
      |         ^~~~~~~~~~~~~~~
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:10789:127: error: expected ')' before numeric constant
10789 |         (*((int*)map_get_and_set((map*)&n, &(string[]){map_set(&m, &(string[]){_SLIT("a")}, &(string[]) { }, &(int[]) { 0 })))2 });
      |         ~                                                                                                                     ^
      |                                                                                                                               )
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:10789:128: error: expected ';' before '}' token
10789 |         (*((int*)map_get_and_set((map*)&n, &(string[]){map_set(&m, &(string[]){_SLIT("a")}, &(string[]) { }, &(int[]) { 0 })))2 });
      |                                                                                                                                ^~
      |                                                                                                                                ;
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c: At top level:
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:10789:130: error: expected identifier or '(' before ')' token
10789 |         (*((int*)map_get_and_set((map*)&n, &(string[]){map_set(&m, &(string[]){_SLIT("a")}, &(string[]) { }, &(int[]) { 0 })))2 });
      |                                                                                                                                  ^
C:\Users\SL\AppData\Local\Temp\v_0\try-maps.188069690999489906.tmp.c:10790:1: error: expected identifier or '(' before '}' token
10790 | }
      | ^
builder error:
==================
C error. This should never happen.
```

--- a/vlib/v/gen/c/index.v
+++ b/vlib/v/gen/c/index.v
@@ -350,7 +350,13 @@ fn (mut g Gen) index_of_map(node ast.IndexExpr, sym ast.TypeSymbol) {
 			g.write('->val')
 		}
 		g.write(', &($key_type_str[]){')
+		old_is_arraymap_set := g.is_arraymap_set
+		old_is_assign_lhs := g.is_assign_lhs
+		g.is_arraymap_set = false
+		g.is_assign_lhs = false
 		g.expr(node.index)
+		g.is_arraymap_set = old_is_arraymap_set
+		g.is_assign_lhs = old_is_assign_lhs
 		g.write('}')
 		if elem_typ.kind == .function {
 			g.write(', &(voidptr[]) { ')

--- a/vlib/v/tests/nested_map_test.v
+++ b/vlib/v/tests/nested_map_test.v
@@ -32,3 +32,15 @@ fn test_map_of_map_to_struct() {
 	foos['zza']['zzb'].name = 'baz'
 	assert foos['zza']['zzb'].name == 'baz'
 }
+
+fn test_map_of_map_key_plus_assign() {
+	m := {
+		'a': 'x'
+	}
+	mut n := {
+		'x': 1
+	}
+	n[m['a']] += 2
+	println(n)
+	assert n['x'] == 3
+}

